File: public/js/core/identity/domain-id.js
// domain-id.js — the domain identity derivation, extracted into its own
// module so both the app (main.js) and anything that needs to VERIFY a
// claimed domain id against a public key (conservation-bridge.js, for
// real transfer-signature checking) use the exact same function. Having
// two independent copies of this — one to derive, one to verify against
// — is exactly the kind of drift that causes silent security bugs; this
// file exists so there is only one.

/**
 * A domain's id: SHA-256 of its wallet's public key, hex-encoded. Full
 * 64 hex characters (256 bits) — NOT truncated. An earlier revision
 * truncated this to 12 hex characters (48 bits) for a shorter display
 * string; that was a real weakness once domain ids became something an
 * attacker could be motivated to find a colliding preimage for (once
 * transfer signatures started binding a signature to "this pubkey
 * hashes to this domain id" — see conservation-bridge.js). 48 bits is
 * within reach of a well-resourced brute-force search; 256 bits is not.
 * Display code that wants a short label may still slice this string for
 * SHOWING it — it must never slice it before using it as the actual
 * identifier a security check depends on.
 *
 * @param {Uint8Array} publicKeyBytes
 * @returns {Promise<string>}
 */
export async function deriveDomainId(publicKeyBytes) {
  const digest = await crypto.subtle.digest('SHA-256', publicKeyBytes);
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

/** A short label for UI display only — never use this as an identifier in a security check. */
export function shortDomainLabel(domainId) {
  return domainId.slice(0, 12);
}

File: rust-core/src/identity/domain_id.rs
use sha2::{Digest, Sha256};

/// A domain's id: SHA-256 of its wallet's public key, hex-encoded — the
/// full 64 hex characters (256 bits), not truncated. Mirror of
/// deriveDomainId() in domain-id.js — see that file's header for why
/// truncation became a real weakness once domain ids started being
/// checked against inside a signature-verification path
/// (conservation_bridge.rs's transfer authorization).
pub fn derive_domain_id(public_key_bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(public_key_bytes);
    hasher.finalize().iter().map(|b| format!("{:02x}", b)).collect()
}

/// A short label for UI display only — never use this as an identifier
/// in a security check.
pub fn short_domain_label(domain_id: &str) -> String {
    domain_id.chars().take(12).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deterministic_same_key_same_id() {
        let key = [7u8; 32];
        assert_eq!(derive_domain_id(&key), derive_domain_id(&key));
    }

    #[test]
    fn full_64_char_hex_not_truncated() {
        let key = [7u8; 32];
        assert_eq!(derive_domain_id(&key).len(), 64);
    }

    #[test]
    fn different_keys_differ() {
        let key1 = [1u8; 32];
        let key2 = [2u8; 32];
        assert_ne!(derive_domain_id(&key1), derive_domain_id(&key2));
    }

    #[test]
    fn short_label_truncates_for_display() {
        let key = [7u8; 32];
        let id = derive_domain_id(&key);
        let label = short_domain_label(&id);
        assert_eq!(label.len(), 12);
        assert!(id.starts_with(&label));
    }
}
